### PR TITLE
add set_target_status

### PR DESCRIPTION
**Pull request by hishamhm**
Among the various options of semantics for `set_target_status`,
I went with a yet simpler one: it merely resets the counters
and sets the status to the given one. That is, it forces a
state momentarily, and restarts health checking, as if the node
was removed and readded with a given "health" status.

This is useful enough for Circuit Breaker functionality: when
a target was marked unhealthy by a passive checker, this can
bring it back up. If it is healthy, it will remain up, if it
is still unhealthy, it will quickly go back down again.

### CODE REVIEW

File: lib/resty/healthcheck.lua [hunk 1]
```diff
@@ -657,6 +622,55 @@ function checker:report_timeout(ip, port, check)
...
+-- @param mode boolean: `true` for healthy, `false` for unhealthy
+-- @return `true` on success, or `nil + error` on failure
+function checker:set_target_status(ip, port, mode)
+  ip   = tostring(assert(ip, "no ip address provided"))
+  port = assert(tonumber(port), "no port number provided")
+  assert(type(mode) == "boolean")
+
+  local health_mode = mode and "healthy" or "unhealthy"
+
+  port = tonumber(port)
```
Comment (Tieske) [inline lib/resty/healthcheck.lua]: this is already covered in the assert above

File: lib/resty/healthcheck.lua [hunk 2]
```diff
@@ -657,6 +622,55 @@ function checker:report_timeout(ip, port, check)
...
+  local target = (self.targets[ip] or EMPTY)[port]
+  if not target then
+    -- sync issue: warn, but return success
+    self:log(WARN, "trying to increment a target that is not in the list: ", ip, ":", port)
```
Comment (Tieske) [inline lib/resty/healthcheck.lua]: copy-pasta: update the error message

File: lib/resty/healthcheck.lua [hunk 3]
```diff
@@ -657,6 +622,55 @@ function checker:report_timeout(ip, port, check)
...
+    local _, err = self.shm:set(oks_key, 0)
+    if err then
+      return nil, err
+    end
+    local _, err = self.shm:set(noks_key, 0)
+    if err then
+      return nil, err
+    end
+
+    local status_key = get_shm_key(self.TARGET_STATUS, ip, port)
```
Comment (Tieske) [inline lib/resty/healthcheck.lua]: nitpick, move it out of the lock, with the other two key-generating calls

File: lib/resty/healthcheck.lua [hunk 4]
```diff
@@ -657,6 +622,55 @@ function checker:report_timeout(ip, port, check)
...
+    self.shm:set(status_key, health_mode == "healthy")
```
Comment (Tieske) [inline lib/resty/healthcheck.lua]: the two `shm:set()` calls above (651-658) have error checking, this one doesn't

File: lib/resty/healthcheck.lua (hunk 2 above)
Comment (hishamhm) [reply lib/resty/healthcheck.lua]: 🙈

**Pull request merged**